If more tags are found during tag definition/declaration goto, let user select which one to use

techee opened the pull request:
If only a single tag is found, just perform the goto. If more tags are found,
show them in the message window.

This patch tries to solve the annoying situation when more tags of the same name exist in the project and the goto just jumps somewhere which is often incorrect. Multiple tag definitions are quite common in OO languages where a virtual method is implemented in several classes or the constructor has the same name as the class. In these situations the current behaviour of the goto is rather useless.

There are two ways how this patch could be further improved:

1. It would be nice if instead of the message window a scintilla popup list would appear with the multiple goto choices. I'm not very familiar with Scintilla so I'm not sure if it's possible (I've found the autocompletion popup lists in Scintilla documentation but I'm not sure if there's some possibility for other kinds of lists).

2. As the tag type I just display the internal TM type name - once we have the table mapping real ctags types to the TM ones, we could use the ctags names (I still plan to implement this feature once I have some time).

b4n commented:
> If only a single tag is found, just perform the goto. If more tags are found,
show them in the message window.

Goal looks sensible, as indeed it often becomes annoying.

> 1. It would be nice if instead of the message window a scintilla popup list would appear with the multiple goto choices. I'm not very familiar with Scintilla so I'm not sure if it's possible (I've found the autocompletion popup lists in Scintilla documentation but I'm not sure if there's some possibility for other kinds of lists).

I guess you could just use a custom popup

> 2. As the tag type I just display the internal TM type name - once we have the table mapping real ctags types to the TM ones, we could use the ctags names (I still plan to implement this feature once I have some time).

Shouldn't this ideally use the same icon/name than in the symbol pane?

ntrel commented:
> If more tags are found, show them in the message window.

This might be annoying if the user wants to keep existing search results there. One alternative would be to have a dedicated *Find Tags* command, outputting to the message window.

techee commented:
>
>
>  Goal looks sensible, as indeed it often becomes annoying.
>
>
>    1. It would be nice if instead of the message window a scintilla popup
>    list would appear with the multiple goto choices. I'm not very familiar
>    with Scintilla so I'm not sure if it's possible (I've found the
>    autocompletion popup lists in Scintilla documentation but I'm not sure if
>    there's some possibility for other kinds of lists).
>
>  I guess you could just use a custom popup
>
This would be probably possible. Do you know how/if it is possible to get
the coordinates of the caret? With Ctrl+click I can just popup the list at
the click position but when used with keybindings, it should probably
happen at the caret position.


>
>    1. As the tag type I just display the internal TM type name - once we
>    have the table mapping real ctags types to the TM ones, we could use the
>    ctags names (I still plan to implement this feature once I have some time).
>
>  Shouldn't this ideally use the same icon/name than in the symbol pane?
>
Sounds good.

techee commented:
> If more tags are found, show them in the message window.
>
> This might be annoying if the user wants to keep existing search results
> there. One alternative would be to have a dedicated *Find Tags* command,
> outputting to the message window.
>
Yes, exactly. This is why I think the popup list of tags is a better
solution.

codebrainz commented:
On 15-01-27 05:15 AM, Jiří Techet wrote:
> This would be probably possible. Do you know how/if it is possible to get
> the coordinates of the caret? With Ctrl+click I can just popup the list at
> the click position but when used with keybindings, it should probably
> happen at the caret position.
>

http://www.scintilla.org/ScintillaDoc.html#SCI_POINTXFROMPOSITION

Used with

http://www.scintilla.org/ScintillaDoc.html#SCI_GETCURRENTPOS

I think ought to do it.

techee commented:
Great, thanks - it's a bit hard to find what you need in the one-page
Scintilla documentation...

techee commented:
I've finally finished this patch. Regarding the popup I decided doing it the lightweight way - I had a look at Scintilla and the autocompletion positioning/resizing code is quite complex. So I just took some reasonable size values for the popup and placed it inside the middle of the window (similarly to the ctrl-tab document switching popup). All the focus grabbing and keyboard-only navigating inside the dialog works (the window mustn't have the POPUP hint otherwise it won't grab the focus). The only things I show in the dialog are

1. Icon corresponding to the filetype (some filetypes don't use any icon in the sidebar so I use some generic tag type -> icon name mapping)
2. Filename (without path to make it shorter)
3. Line number of the tag

sorted alphabetically by name.

After playing with the result for a while I realized that having to go through the list is sometimes annoying and there are situations where we can guess the best tag and this tag can be placed at the first position (shown in bold). The "best" tag must be either

1. from the current file
2. from an open document
3. within the current file's directory
4. within the current project base directory

techee commented:
I've just rebased this patch on top of master with some minor fixes.

I'd really like to see this patch in 1.26 because the jump to the arbitrary tag is super-annoying.

elextr commented:
Ping (on behalf of @techee )

Like the idea.

techee commented:
@b4n Not really a zombie as it hasn't been reviewed yet but I'd love to get this in (or in other words I hate getting to the wrong location when going to tag definition/declaration). So if you feel terribly bored, longing to review some patch, this is the one :-)

b4n commented:
Hum, I don't really like the popup, esp. because it can popup far from the pointer or whatever the trigger was (and it be something else?).  Also, the popup sizing is incorrect (at least with my theme, so it's incorrect in general, yet maybe happen to be good in your case).  This is fixable correctly (not like in Scintilla), but not really trivial.

Anyway, I wanted to check if it was easy to implement using AUTOCLISTs, and realized Scintilla had USERLISTs, which sounded quite nice :)
Here is a sample of how it might look like (yet it has 2 unresolved issues, one hard and one simple):
```c
static void on_goto_popup_sci_notify(GtkWidget *sci, gint scn, SCNotification *nt, gpointer data)
{
	switch (nt->nmhdr.code)
	{
		case SCN_USERLISTSELECTION:
			if (nt->listType == 42)
			{
				gchar *name = g_strdup(nt->text);
				gchar *line = strrchr(name, ':');
				GeanyDocument *old_doc = document_get_current();
				GeanyDocument *new_doc;

				if (line)
				{
					*line = 0;
					line++;
				}

				new_doc = document_find_by_real_path(name);
				if (!new_doc)
					new_doc = document_open_file(name, FALSE, NULL, NULL);

				if (new_doc)
					navqueue_goto_line(old_doc, new_doc, line ? atoi(line) : 0);

				g_free(name);
			}
			/* fallthrough */
		case SCN_AUTOCCANCELLED:
			g_signal_handlers_disconnect_by_func(sci, on_goto_popup_sci_notify, data);
			break;
	}
}


static void show_goto_popup(GeanyDocument *doc, GPtrArray *tags, gboolean have_best)
{
	TMTag *tmtag;
	guint i;
	GString *words;

	if (tags->len < 1)
		return;

	words = g_string_new(NULL);

	foreach_ptr_array(tmtag, i, tags)
	{
		if (words->len > 0)
			g_string_append_c(words, '\n');

		/* FIXME: how to show something and get associated data? */
		g_string_append_printf(words, "%s:%lu", tmtag->file->file_name, tmtag->line);

		/* FIXME: better images */
		if (!EMPTY(tmtag->arglist))
			g_string_append(words, "?2");
		else
			g_string_append(words, "?1");
	}

	g_signal_connect(doc->editor->sci, "sci-notify", G_CALLBACK(on_goto_popup_sci_notify), NULL);
	scintilla_send_message(doc->editor->sci, SCI_AUTOCCANCEL, 0, 0);
	scintilla_send_message(doc->editor->sci, SCI_USERLISTSHOW, 42, (sptr_t) words->str);
	g_string_free(words, TRUE);
}
```
It's quick'n dirty, and we might rather want to register that in the editor handler or somewhere -- although this doesn't feel *so* bad in the end, as it's contained and temporary.  The real question here is the first *FIXME in the code, e.g. whether it's possible to store some additional data with the displayed entry.  Otherwise, we either need something like I did, or a table mapping the name to the actual tag… meh.

---

Apart that, I don't know, the theory is probably fine, but to really have an opinion I'd need to use it for some time and see in the "long" run.
And there's an issue when merging this in current master, conflicting signature for `tm_workspace_find()`.  And an int should be unsigned :)
```diff
diff --git a/src/symbols.c b/src/symbols.c
index ed71004..7e429ec 100644
--- a/src/symbols.c
+++ b/src/symbols.c
@@ -1990,7 +1990,7 @@ static void show_goto_popup(GPtrArray *tags, gboolean have_best)
     GtkTreeModel *model;
     GtkTreeIter iter;
     TMTag *tmtag;
-    gint i;
+    guint i;
 
     if (!tag_goto_popup)
         create_goto_popup();
@@ -2094,7 +2094,7 @@ static gboolean goto_tag(const gchar *name, gboolean definition)
 
     /* goto tag definition: all except prototypes / forward declarations / externs */
     type = (definition) ? tm_tag_max_t - forward_types : forward_types;
-    all_tags = tm_workspace_find(name, type, NULL, FALSE, old_doc->file_type->lang);
+    all_tags = tm_workspace_find(name, NULL, type, NULL, old_doc->file_type->lang);
 
     /* get rid of global tags */
     workspace_tags = g_ptr_array_new();
```
(the `tm_workspace_find()` might not be correctly fixed, I did that quickly and I'm not yet familiar with the new API)

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+
+	new_doc = document_find_by_real_path(tag->file->file_name);
+	if (!new_doc)
+		new_doc = document_open_file(tag->file->file_name, FALSE, NULL, NULL);
```
won't that change what is the "current" document below, and then lead to a weird sequence in the location history?

b4n commented on src/symbols.c:
```diff
@@ -1953,10 +2158,34 @@ static gboolean goto_tag(const gchar *name, gboolean definition)
 
-		if (navqueue_goto_line(old_doc, new_doc, tmtag->line))
-			return TRUE;
+		if (!swapped && navqueue_goto_line(old_doc, new_doc, tmtag->line))
```
couldn't it simply use `found`, and drop the whole `swapped` variable?

b4n commented on src/symbols.c:
```diff
@@ -1953,10 +2158,34 @@ static gboolean goto_tag(const gchar *name, gboolean definition)
+			if (tag != best_tag)
+				g_ptr_array_add(tags, tag);
+		}
+		show_goto_popup(tags, best_tag != 0);
```
`0` is not a pointer, should rather be `NULL`

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+
+static TMTag *find_best_goto_tag(GPtrArray *tags)
+{
+	GeanyDocument *doc = document_get_current();
```
I'd rather pass this as an argument, lowering the environmental factor of this function

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+}
+
+
+static gint compare_tags(gconstpointer ptr1, gconstpointer ptr2)
```
not directly clear how it sorts the tags, so maybe either rename the function to be explicit, or at least add a small comment so one doesn't have to read the body just to get "ah, k, just sorts by basename, then line" :)

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+	GtkTreeViewColumn *column;
+	GtkCellRenderer *renderer;
+
+	tag_goto_popup = g_object_new(GTK_TYPE_WINDOW, "type", GTK_WINDOW_TOPLEVEL, NULL);
```
note: the default constructor, [`gtk_window_new()`](https://developer.gnome.org/gtk2/stable/GtkWindow.html#gtk-window-new) has just those parameters, why not use it :)

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+	gtk_window_set_decorated(GTK_WINDOW(tag_goto_popup), FALSE);
+	gtk_window_set_transient_for(GTK_WINDOW(tag_goto_popup), GTK_WINDOW(main_widgets.window));
+	gtk_window_set_destroy_with_parent(GTK_WINDOW(tag_goto_popup), TRUE);
+	gtk_window_set_position(GTK_WINDOW(tag_goto_popup), GTK_WIN_POS_CENTER_ON_PARENT);
```
…or at least pass some of those as gobject construction props

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+	gtk_window_set_transient_for(GTK_WINDOW(tag_goto_popup), GTK_WINDOW(main_widgets.window));
+	gtk_window_set_destroy_with_parent(GTK_WINDOW(tag_goto_popup), TRUE);
+	gtk_window_set_position(GTK_WINDOW(tag_goto_popup), GTK_WIN_POS_CENTER_ON_PARENT);
+	gtk_widget_set_size_request(tag_goto_popup, 250, 150);
```
meh

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+			break;
+	}
+	return "classviewer-struct";
+}
```
FIXME: use the same icons as the symbols tree

b4n commented on src/symbols.c:
```diff
@@ -138,6 +148,9 @@ static struct
 symbol_menu;
 
+static GtkWidget *tag_goto_popup = NULL;
+static GtkWidget *tag_goto_tree_view = NULL;
```
meh, is it that important to cache the popup?

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+		navqueue_goto_line(old_doc, new_doc, tag->line);
+	}
+
+	gtk_widget_hide(tag_goto_popup);
```
the whole store should probably be cleared when hiding, not to keep TMTags around for no reason.  note: destroying the popup would take care of this.

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+
+static gboolean on_key_pressed(GtkWidget *widget, GdkEventKey *event, gpointer user_data)
+{
+	guint state = event->state & gtk_accelerator_get_default_mod_mask();
```
should probably use `keybindings_get_modifiers()`

b4n commented on src/symbols.c:
```diff
@@ -54,6 +54,8 @@
 #include "tm_tag.h"
 #include "ui_utils.h"
 #include "utils.h"
+#include "msgwindow.h"
```
leftover?

b4n commented:
I think I'm finished spamming this PR for the moment :)

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+			text = g_strdup_printf("%s: %lu", fname, tmtag->line);
+
+		gtk_list_store_insert_with_values(GTK_LIST_STORE(model), &iter, -1,
+				PIXBUF_COLUMN, get_tag_icon(get_tag_class(tmtag)),
```
this leaks the pixbuf I think

b4n replied:
I suggest something like this https://github.com/b4n/geany/commit/5533a5031780867a97e157d200dba8c6f8bfb569

b4n commented on src/symbols.c:
```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+		if (first && have_best)
+			text = g_strdup_printf("<b>%s: %lu</b>", fname, tmtag->line);
+		else
+			text = g_strdup_printf("%s: %lu", fname, tmtag->line);
```
text should be escaped for markup (g_markup_printf_escaped())

b4n replied:
https://github.com/b4n/geany/commit/98194c007d98f22cd45ebe03b56ab5fbd3c86e40

b4n commented:
Current situation (and USERLIST one suggested above too) has bad usability with the mouse (ctrl+click) as it requires a double click.  As there is a primary mouse-based trigger, I think it's important to fix this.
(It probably could be nice to also directly select on single click from Scintilla's AUTOCLIST, but as wa don't have any mouse-based triggers currently it's not that important)

Anyway I tried to re-implement this using a `GtkMenu` and `gtk_menu_popup()`, and it had a subtlety (because we don't get the proper `GdkEvent` to feed `gtk_menu_popup()`) but otherwise was pretty straightforward, simpler (all popup logic is left out to `GtkMenu`), and behaves better regarding input, as click selects, and keyboard navigation works fine.  As it isn't likely there would be an endless list of possibilities, and anyway if there was the UI here wouldn't be really usable, I don't think there is a problem with putting elements in a menu (which doesn't scroll before filling the screen).
WIP branch: https://github.com/b4n/geany/commits/wip/techee/multi_tag_goto%2Bpopup-menu-wip

---

> Regarding the popup I decided doing it the lightweight way - I had a look at Scintilla and the autocompletion positioning/resizing code is quite complex. So I just took some reasonable size values for the popup and placed it inside the middle of the window (similarly to the ctrl-tab document switching popup).

The complexity from Scintilla's popup placement is the will to align the text column exactly to the cursor, so that the autocompletion suggestion is perfectly vertically aligned with the buffer's text.  This means location the exact X coordinate of the second row's cell renderer, which isn't that trivial.
In my above-mentioned menu-based popup I implemented [a straightforward version of the popup placement](https://github.com/b4n/geany/blob/wip/techee/multi_tag_goto%2Bpopup-menu-wip/src/symbols.c#L1884) that simply places the top-left of the popup below the cursor.  I believe this to be good enough in this case, as anyway the item text has nothing to do with the buffer's content.

techee commented:
@b4n Big thanks for having a look at the patch! I'll check your comments and your implementation if the popup and will update the patch (probably tomorrow).

techee commented:
@b4n I tested your patches and things look pretty great. I have also addressed your comments and pushed the changes here:

https://github.com/techee/geany/tree/multi_tag_goto_2

The only thing I didn't like much was that on Ctrl+click the first item isn't selected - I think even when using Ctrl+click one should be able to just press enter to go to the first, suggested, tag. I removed the special case in the penultimate patch in the branch above.

The second problem is that the push_in parameter causes this:

![screenshot_2016-02-18_21-54-31](https://cloud.githubusercontent.com/assets/713965/13158613/00b47eda-d68e-11e5-8107-45f4425ff8cd.png)

I think I prefer push_in = FALSE, which I did in the last patch and which, while not ideal, looks this way:

![screenshot_2016-02-18_21-53-13](https://cloud.githubusercontent.com/assets/713965/13158676/5320ac48-d68e-11e5-8255-708f13caa0ab.png)

(It would be best to have the popup over the whole screen like in the first screenshot but filled with all the values without the empty space.)

b4n commented:
> I think I prefer push_in = FALSE, which I did in the last patch and which, while not ideal, looks this way:

Well, it's bad that `push_in=TRUE` doesn't scroll right away (or maybe not, so the most likely element is close to the mouse), but at least the menu sizing is decent.  with FALSE, try and place the popup location close enough to the bottom and you'll see it's so small you actually can't see the items.
Also, with `push_in=TRUE`, the initial state is not great, but you can scroll it and it gets it right then.

b4n commented:
Hum, actually on GTK 3.18 `push_in=FALSE` is slightly better for our case, because it pushes the menu (yes, it does), but just moving it instead of scrolling it, so the whole menu is visible.

b4n commented:
If you want a more sensible popup placement, it's doable:
```C

/* positions a popup at the caret from the ScintillaObject in @p data */
static void goto_popup_position_func(GtkMenu *menu, gint *x, gint *y, gboolean *push_in, gpointer data)
{
	ScintillaObject *sci = data;
	GdkWindow *window = gtk_widget_get_window(GTK_WIDGET(sci));
	gint pos = sci_get_current_position(sci);
	gint line = sci_get_line_from_position(sci, pos);
	gint line_height = scintilla_send_message(sci, SCI_TEXTHEIGHT, line, 0);
	gint pos_x = scintilla_send_message(sci, SCI_POINTXFROMPOSITION, 0, pos);
	gint pos_y = scintilla_send_message(sci, SCI_POINTYFROMPOSITION, 0, pos);
	GdkScreen *screen = gtk_widget_get_screen(GTK_WIDGET(menu));
	gint monitor_num;
	GdkRectangle monitor;
	GtkRequisition req;

	gdk_window_get_origin(window, x, y);
	monitor_num = gdk_screen_get_monitor_at_point(screen, *x + pos_x, *y + pos_y);

#if GTK_CHECK_VERSION(3, 0, 0)
	gtk_widget_get_preferred_size(GTK_WIDGET(menu), NULL, &req);
#else
	gtk_widget_size_request(GTK_WIDGET(menu), &req);
#endif

#if GTK_CHECK_VERSION(3, 4, 0)
	gdk_screen_get_monitor_workarea(screen, monitor_num, &monitor);
#else
	gdk_screen_get_monitor_geometry(screen, monitor_num, &monitor);
#endif

	/* put at the X position, but within the monitor */
	if (gtk_widget_get_direction(GTK_WIDGET(menu)) == GTK_TEXT_DIR_RTL)
		*x = MAX(monitor.x, *x + pos_x - req.width);
	else
	{
		if (*x + pos_x + req.width <= monitor.x + monitor.width)
			*x = MAX(monitor.x, *x + pos_x);
		else
			*x = monitor.x + MAX(0, monitor.width - req.width);
	}

	/* try to put, in order:
	 * 1. below the Y position, under the line
	 * 2. above the Y position
	 * 3. within the monitor */
	if (*y + pos_y + line_height + req.height <= monitor.y + monitor.height)
		*y = MAX(monitor.y, *y + pos_y + line_height);
	else if (*y + pos_y - req.height >= monitor.y)
		*y = *y + pos_y - req.height;
	else
		*y = monitor.y + MAX(0, monitor.height - req.height);

	*push_in = FALSE;
}
```
It's a fair bit more complex, but more accurate without getting into too much details.  Basically, it puts the popup below if it fits, otherwise tries above, else puts it as far low as possible within the monitor.  For the X position, puts it on the natural side, but inside the monitor.

It's not exactly what "natural" is for GTK, which would put preferably a corner next to the popup point, and otherwise chose the side with the most space, but implementing this is a little more work for not much gain.

b4n commented:
Updated my WIP branch (https://github.com/b4n/geany/commits/wip/techee/multi_tag_goto%2Bpopup-menu-wip) with your changes and the mentioned popup placement function.

elextr commented:
@b4n why not use screen width and height so geany can spread across multiple monitors if the screen is?

b4n commented:
@elextr meh… do you really want a menu spanning several monitors?  Also, it would only make some sense with vertical spanning, as the menu is not likely to be wider than the monitor, but rather (possibly) taller.  And even then, I think it's best to at least consider the monitor not to put a half of an entry in the next monitor in the screen -- so IMO the only possible moment where you'd really consider spanning multiple monitors is if you can't fit in the current one, not before.
Finally, all this is pretty theoretical when it comes to a list taller than the monitor, if you really get to this in real situation you'll have other worries I'm afraid :)

elextr commented:
@b4n, if the user has geany spread across two monitors then clearly they are happy with that, so why not make the menu do the same?  Reduces the menu jumping up and down depending how long it is.

Just suggesting, seemed like only a change of calls for the size, but anyway if its too complex forget it.

b4n commented:
Not considering monitors is easy, sure.  Considering them only when Geany doesn't span several monitors sounds a lot more complex though, would require checking which monitors Geany spans on, compute the bounding box and use that.  And in any other case, I really think trying to fit in the monitor seem more sensible to me.

Anyway, if someone really wants a screen-level logic, why not, but I'd like it to be smart enough not to mess up the most common situation where your Geany is on a single monitor and you want to keep it that way :)

techee commented:
@b4n Your patch works pretty great - there was just a minor problem I tried to solve in the last commit of

https://github.com/techee/geany/tree/multi_tag_goto_2

techee commented:
@b4n I made one more update here

https://github.com/techee/geany/tree/multi_tag_goto_2

I believe it works quite well now.

techee commented:
@b4n How should I proceed with the patch? I think the final version after all your improvements works really well - is it OK to squash all the patches together with big thanks to you? (LOC-wise the patch will be more yours than mine now.) Or should I keep them separate? (At least the small patches should be squashed I think.)

b4n commented:
@techee do as you see fit, I don't mind losing some of the credit or anything ^^
I guess keeping some of the iterations might be interesting, or at least their reasoning when it was non-obvious (e.g. when there was a large-ish comment :)). But yeah some commits are mostly noise so they could get squashed ("small fixes" of mine comes to mind).  However if it's hard to get something nice, don't bother too much and we can merge as-is too (I like the history a bit less, but well, not a biggie either).

techee commented:
@b4n I repushed the patch with the various fixes and improvements - I squashed only the small things and left the bigger ones (or ones from different authors) as separate commits.

techee commented:
@b4n Fantastic! Thanks for merging.
